File: 백준/Gold/1167. 트리의 지름/트리의 지름.cc
#include <iostream>
#include <vector>

using namespace std;

const int MX = 100002;
vector<pair<int, int>> vec[MX];
int v;
long long dist[MX];

void dfs(int cur, int par){
    for(auto p : vec[cur]){
        auto [child, d] = p;
        if(child == par) continue;
        dist[child] = d;
        dist[child] += dist[cur]; 
        dfs(child, cur);
    }
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> v;
    
    for(int i = 0; i < v; i++){
        int a , b , d;
        cin >> a;
   
        while(1){
            cin >> b;
            if(b == -1) break;
            cin >> d;
            vec[a].push_back({b, d});
        }
    }
    
    
    int r = 1;
    dfs(r, -1);
    int mx = -1;
    for(int i = 1; i <= v; i++){
        if(mx <= dist[i]){
            mx = dist[i];
            r = i;
        }
        dist[i] = 0;
    }
    dfs(r, -1);
     for(int i = 1; i <= v; i++){
        if(mx <= dist[i]){
            mx = dist[i];
            r = i;
        }
    }
   
    cout << mx;
    return 0;
}

File: 백준/Gold/1967. 트리의 지름/트리의 지름.java
import java.io.FileInputStream;
import java.io.FileNotFoundException;

import java.util.ArrayList;

import java.util.List;
import java.util.Scanner;

class Pair {

    public int x;
    public int y;

    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }


}

public class Main {

    static List<Pair>[] vec = new List[10002];

    static int dists[] = new int[10002];

    public static void dfs(int cur, int par) {
        for (Pair pair : vec[cur]) {
            int child = pair.x;
            int dist = pair.y;
            if (child == par) {
                continue;
            }
            dists[child] = dist + dists[cur];
            dfs(child, cur);
        }
    }

    public static void main(String[] args) throws FileNotFoundException {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        for (int i = 0; i <= n; i++) {
            vec[i] = new ArrayList<>();
        }

        for (int i = 0; i < n - 1; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            int c = sc.nextInt();

            vec[a].add(new Pair(b, c));
            vec[b].add(new Pair(a, c));
        }

        StringBuilder sb = new StringBuilder();
        int root = 1;
        dfs(root, 0);
        int mx = -1;
        for (int i = 1; i <= n; i++) {
            if (mx < dists[i]) {
                mx = dists[i];
                root = i;
            }
            dists[i] = 0;
        }

        dfs(root, 0);

        for (int i = 1; i <= n; i++) {
            if (mx < dists[i]) {
                mx = dists[i];
            }
        }

        System.out.println(mx);
    }
}